proposals/0000-dictionary-key-and-value-collections.md

```diff
@@ -0,0 +1,176 @@
+
+```swift
+// Fast, not slow
+if dict.keys.contains("one") {
+    // ...
+}
+```
+
+As a mutable collection, `values` enables modification without copies or clumsy code:
+
+```swift
+if let i = dict.index(forKey: "one") {
+    dict.values[i].append(1)  // no copy here
+} else {
+    dict["one"] = [1]
+}
+```
+
+Both the `keys` and `values` collections share the same index type as `Dictionary`. This allows the above sample to be rewritten as:
+
+```swift
+// Using `dict.keys.index(of:)`
+if let i = dict.keys.index(of: "one") {
+    dict.values[i].append(1)
+} else {
+    dict["one"] = [1]
+}
+```
+
+## Detailed design
+
+* The standard library introduces two new collection types: `Dictionary.Keys` and `Dictionary.Values`.
+* A `Dictionary`'s `keys` and `values` properties change from `LazyMapCollection` to these new types. 
+* The new collection types are not directly constructable. They are presented only as views into a dictionary.
+
+```swift
+struct Dictionary<Key: Hashable, Value>: ... {
+    /// A collection view of a dictionary's keys.
+    struct Keys: Collection {
+        subscript(i: Index) -> Key { get }
+        // Other `Collection` requirements
+    }
+
+    /// A mutable collection view of a dictionary's values.
+    struct Values: MutableCollection {
+        subscript(i: Index) -> Value { get set }
+        // Other `Collection` requirements
+    }
+    
+    var keys: Keys { get }
+    var values: Values { get set }    
+    // Remaining Dictionary declarations
+}
+```
+
+## Impact on existing code
+
+The performance improvements of using the new `Dictionary.Keys` type and the mutability of the `Dictionary.Values` collection are both additive in nature.
+
+Most uses of these properties are transitory in nature. Adopting this proposal should not produce a major impact on existing code. The only impact on existing code exists where a program explicitly specifies the type of a dictionary's `keys` or `values` property. In those cases, the fix would be to change the specified type.
+
+
+## Alternatives considered
+
+1. Add additional compiler features that manage mutation through existing key-based subscripting without the copy-on-write problems of the current implementation. This could potentially be handled by upcoming changes to copy-on-write semantics and/or inout access. 
+
+2. Provide new APIs for updating dictionary values with a default value, eliminating the double-lookup for a missing key. The approach outlined in this proposal provides a way to remove one kind of double-lookup (mutating a value that exists) but doesn't eliminate all of them (in particular, checking for the existence of a key before adding).
+
+   These could be written in a variety of ways:
+    
+    ```swift
+    // Using a 'SearchState' type to remember key position
+    dict.entries["one"]
+        .withDefault([])
+        .append(1)
+    
+    // Using a two-argument subscript
+    dict["one", withDefault: []].append(1)
```

Comment:
I don't think there was a fatal blow; I was just trying to keep the proposal focused on one thing. I can add the default value subscript so it can be part of the discussion—I think the core team has been judicious with its line-item veto when necessary.